readme.md
### Preliminary note:

This is a maintenance/fix/improvement release of the (apparantly) abandoned [Nudist](https://www.curseforge.com/wow/addons/nudist) addon by tekkub. 

Some of the code is unchanged from the original. All images for the iconic Gnome button are unchanged. So, of course credits to tekkub.

If you are the author of the original addon, and don’t like to see this fork of your addon posted here or anywhere, just leave a note in the Issues thread.


## What this addon does:

Clicking the Gnome button on your char frame or entering “/nudist” in the console (or via macro) strips your char off of all items that would suffer from death (durability-wise).

Doing the same again re-equips you to exactly the same state you were at when you clicked the button first (or when you entered the command).
This is the main point of this addon: It re-equips you to the state of when you clicked the button, completely independent of any Blizz equip sets or BtWLoadouts equip sets or such. 

What for? I can create a “Nude” set?

A “nude” equip set works of course fine for getting nude. But to de-nude (reequip) you would have to call your equip sets, and often these are not only one but some sets in a row. E.g., with a base equip set for the class + set for the spec + set for the weapons + set for the legendary you would have to activate 4 equip sets to get back to the state where you have been before nuding. 

Nudist solves this issue by one-click re-equipping exactly what you had at the point in time when you first clicked Nudist. It does not know or care about your equip sets or BtWLoadouts, i.e. the addon memorizes your actual equips and restores them. (Thus, it restores equally fine if your equipment wasn’t saved in a set at all.)

## Notes on changes since the “old” version by tekkub:

With the last version of the Nudist addon dating back to 2011 (Cataclysm), I took the liberty to refresh it a bit and make it work with SL.

To my surprise, the last (WoW 4) version of the addon still “worked” with Shadowlands, but there was one major issue:

Trying to re-equip two items that have the same item link (i.e. same item ID and same enchants, may happen with dual-wielded daggers and such) failed without warning. (You ended up with just _one_ weapon, either in slot 16 or in slot 17.)

I solved (hopefully 100%) this problem by restructuring the way the tables are built and introducing an additional check to see if all weapon slots are equipped. If not, we do `UseContainerItem(bag, bagslot)` on the empty slot, after a short delay. `UseContainerItem` is more expensive than `EquipItemByName` (which is used for the “uncomplicated” re-equips) but seems to be the only way to get those dupe-ID items back into their slots correctly.

As a new feature, now you have also the option to unequip _all_ slots. For this, hold down any modifier key while clicking the button or executing the command or macro. (Useful if you have to strip everything to reduce damage output for low-level achievements etc.)


changelog.md
#### 1.2.1 (2022-11-24)
- Works now with 10.0.2
  - Sneaky Blizz guys changed the offset of the bag numbers from 19 to 30!
- Minor optimizations
- toc: updated client version number

#### 1.2.0 (2022-09-15)
- Button status now closer tied to real equipped/nude status
- Improved logic of the equal-weapon-ID reequip, plus additional check
- Added status and failure messages
- Fixed full nude bug
- Minor optimizations

#### 1.1.0 (2022-09-10)
- Disabled also re-equipping if in combat lockdown (instead of only unequipping)
- _New feature_: If a modifier key is down while clicking the button (or launching the command/macro), then all slots will be unequipped.
  - This can be useful in situations where you have to strip everything, in order to reduce damage output (e.g. for old instance achievements where the boss must not be one-shot).
- toc: update for 9.2.7

#### 1.0.0 (2022-08-13)
- Equal item links are now re-equipped correctly (e.g. dual-wielding weapons with same ID and same enchant ID)
- Removed some legacy stuff from code and toc
- Everything up to date for 9.2.7

***
Release notes of the original addon (as by tekkub, https://www.curseforge.com/members/tekkub/projects):

4.1.0.6-Release
 - Tweak position to not overlap the new tabs
 - Move the original images off so as not to confuse wow

4.0.0.5-Release
 - Fix gnome's parent so it only shows on the base char panel

3.2.0.4-Release
 - Don't allow re-equipping in combat
 - Use link and slot for reequipping, to ensure the items go back where they were

3.1.0.3-Release
 - Fix AL condition for 3.1

3.0.1.2-Release
 - Remove unneeded code that's erroring in Wrath



Nudist.lua

local slots, freeslots, items, allslots = {1, 3, 5, 6, 7, 8, 9, 10, 16, 17}, {}, {}, {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 19}


local function GetEmpties()
	for i=0,4 do freeslots[i] = 0 end
	for bag=0,4 do
		for slot=1,GetContainerNumSlots(bag) do
			if not GetContainerItemInfo(bag, slot) then freeslots[bag] = freeslots[bag] + 1 end
		end
	end
end


local function GetNextEmpty()
	for i=0,4 do
		if freeslots[i] > 0 then
			freeslots[i] = freeslots[i] - 1
			return i
		end
	end
end

local function EquipItemByBagSlot(item, slot)
	for bag = 0, 4 do
		local max = GetContainerNumSlots(bag)
		for bagslot = 1, max do
			if GetContainerItemLink(bag, bagslot) == item then
				UseContainerItem(bag, bagslot)
				print("Reequipped missing item via 'UseContainerItem': Slot", slot, item)
				return
			end
		end
	end
end

---------------------------------
--      Char frame button      --
---------------------------------

local butt = CreateFrame("Button", nil, PaperDollFrame)
butt:SetFrameStrata("DIALOG")
butt:SetPoint("BOTTOMRIGHT", CharacterFrame, "TOPRIGHT", -60, -30)
butt:SetWidth(64) butt:SetHeight(64)

-- Textures --
butt:SetNormalTexture("Interface\\Addons\\Nudist\\clothed")
butt:SetPushedTexture("Interface\\Addons\\Nudist\\nude")

-- Tooltip bits
--~ butt:SetScript("OnEnter", ShowTooltip)
--~ butt:SetScript("OnLeave", HideTooltip)


local function handler()
	
	if IsModifierKeyDown() then slots = allslots end 
	if CursorHasItem() then ClearCursor() end
	if InCombatLockdown() then return end
	
	if #items > 0 then

		for s, i in pairs(items) do
			EquipItemByName(i, s)
		end

		C_Timer.After(0.8, function()
			for s = 16, 17 do -- Check only weapon slots [^1]
				if not GetInventoryItemLink("player", s) then
					EquipItemByBagSlot(items[s], s)
				end
			end
			table.wipe(items)
		end)

		butt:SetNormalTexture("Interface\\Addons\\Nudist\\clothed")
		butt:SetPushedTexture("Interface\\Addons\\Nudist\\nude")

	else

		butt:SetNormalTexture("Interface\\Addons\\Nudist\\nude")
		butt:SetPushedTexture("Interface\\Addons\\Nudist\\clothed")
		GetEmpties()

		for _,i in ipairs(slots) do
			local bag = GetNextEmpty()
			if not bag then return end

			local item = GetInventoryItemLink("player", i)
			if item then
				table.insert(items, i, item)
				PickupInventoryItem(i)
				if bag == 0 then PutItemInBackpack() else PutItemInBag(bag + 19) end
			end

		end
	end
end

SLASH_NUDIST1 = "/nudist"
SlashCmdList["NUDIST"] = handler
butt:SetScript("OnClick", handler)



--[=[

Notes
-----

[^1]:

If we have two of the same items, and with the same enchants, equipped, then they have probably an identical item link. Usually this happens only with the weapon slots (16 and 17).

In that case – for some reason – we cannot equip both items with EquipItemByName(itemlink, slot): One item will be equipped correctly, the other one will not.

As a workaround, we have to equip the second one of the identical items via bag slot: we search all bag slots for a matching item link and then equip the item via UseContainerItem(bag, bagslot).

We could use this method right from the start for slots 16 and 17. But I think it's more economic to equip normally, and check afterwards if one of the originally equipped slots is empty. Then we use the workaround for the missing item only. Currently the check is limited to slots 16 and 17, because _I think_ it is impossible to have identical items in the other slots.
The C_Timer is necessary, otherwise, if the check comes to quickly, we get all slots returned as not equipped. 1 second seems to be safe, probably we can go a bit faster (0.25s though is too short).

To check _all_ slots for missing items, use this:

-- 			for s, i in pairs(items) do -- Check all slots [^1]
-- 				if not GetInventoryItemLink("player", s) then
-- 					EquipItemByBagSlot(i, s)
-- 				end
-- 			end


]=]


main.lua
local function dp(...) -- debugprint
	-- return
	print('Nudist Debug:', ...)
end

--[[
This has changed in DF! (was 19).
https://www.wowinterface.com/forums/showthread.php?s=285e4b422a2f9c7ae7954929a74a3ac7&p=341652
https://github.com/Ketho/wow-ui-source-df/search?q=CONTAINER_BAG_OFFSET
]]
local CONTAINER_BAG_OFFSET = 30

local SENSITIVE_SLOTS, ALL_SLOTS, items, freeslots =
	{ 1, 3, 5, 6, 7, 8, 9, 10, 16, 17 },
	{ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 19 }, -- 18 is no longer used in Retail
	{},
	{}

local MSG_PREFIX, MSG_NUDE, MSG_FULL_NUDE, MSG_CLOTHED, MSG_FAILURE =
	'Nudist:',
	'Almost nude now!',
	'Really nude now!',
	'Equipped again.',
	'|cffFF0000Something went wrong while trying to reequip your weapon slots. Check your equipment!'

local function status_msg(...) print(MSG_PREFIX, ...) end

local btn = CreateFrame('Button', nil, PaperDollFrame)
btn:SetFrameStrata 'DIALOG'
btn:SetPoint('BOTTOMRIGHT', CharacterFrame, 'TOPRIGHT', -60, -30)
btn:SetWidth(64)
btn:SetHeight(64)
btn:SetNormalTexture 'Interface/Addons/Nudist-Again/clothed'
btn:SetPushedTexture 'Interface/Addons/Nudist-Again/nude'

local function get_empties()
	for bag = 0, 4 do
		freeslots[bag] = C_Container.GetContainerNumFreeSlots(bag)
	end
end

local function get_next_empty()
	for i = 0, 4 do
		if freeslots[i] > 0 then
			freeslots[i] = freeslots[i] - 1
			return i
		end
	end
end

local function equip_item_by_bagslot(item, slot)
	for bag = 0, 4 do
		local max = C_Container.GetContainerNumSlots(bag)
		for bagslot = 1, max do
			if C_Container.GetContainerItemLink(bag, bagslot) == item then
				C_Container.UseContainerItem(bag, bagslot)
				status_msg("Reequipped missing item via 'UseContainerItem': Slot", slot, item) -- debug
				return
			end
		end
	end
end

local function unequip_done(m)
	btn:SetNormalTexture 'Interface/Addons/Nudist-Again/nude'
	btn:SetPushedTexture 'Interface/Addons/Nudist-Again/clothed'
	status_msg(m)
end

local function reequip_done(m)
	table.wipe(items)
	btn:SetNormalTexture 'Interface/Addons/Nudist-Again/clothed'
	btn:SetPushedTexture 'Interface/Addons/Nudist-Again/nude'
	status_msg(m)
end



local function handler()
	--[[ If in combat, we do nothing, also not trying to reequip, bc chances are
	very high that we get a Protected Function error and end up with an
	equipment item on the cursor, which is disturbing. ]]
	if InCombatLockdown() then return end

	if CursorHasItem() then ClearCursor() end

	if #items > 0 then
		for s, i in pairs(items) do
			EquipItemByName(i, s)
		end

		local msg, verify = MSG_CLOTHED, false

		C_Timer.After(0.8, function()
			for s = 16, 17 do -- Check only weapon slots [^1]
				if items[s] and not GetInventoryItemLink('player', s) then
					equip_item_by_bagslot(items[s], s)
					verify = true
				end
			end
			if not verify then
				reequip_done(msg)
			else -- Doublecheck if the 2nd slot 16/17 attempt was successful
				C_Timer.After(0.8, function()
					for s = 16, 17 do
						if items[s] and not GetInventoryItemLink('player', s) then
							msg = MSG_FAILURE
							break
						end
					end
					reequip_done(msg)
				end)
			end
		end)
	else
		get_empties()

		local slots, msg = SENSITIVE_SLOTS, MSG_NUDE

		if IsModifierKeyDown() then
			slots, msg = ALL_SLOTS, MSG_FULL_NUDE
		end

		for _, i in ipairs(slots) do
			local bag = get_next_empty()
			if not bag then return end

			local item = GetInventoryItemLink('player', i)
			if item then
				table.insert(items, i, item)
				PickupInventoryItem(i)
				if bag == 0 then
					PutItemInBackpack()
				else
					PutItemInBag(bag + CONTAINER_BAG_OFFSET)
				end
			end
		end
		unequip_done(msg)
	end
end

SLASH_NUDIST1 = '/nudist'
SlashCmdList['NUDIST'] = handler
btn:SetScript('OnClick', handler)



--[=[

Notes
-----

[^1]:

If we have two of the same items, and with the same enchants, equipped, then they have probably an identical item link. Usually this happens only with the weapon slots (16 and 17).

In that case – for some reason – we cannot equip both items with EquipItemByName(itemlink, slot): One item will be equipped correctly, the other one will not.

As a workaround, we have to equip the second one of the identical items via bag slot: we search all bag slots for a matching item link and then equip the item via UseContainerItem(bag, bagslot).

We could use this method right from the start for slots 16 and 17. But I think it's more economic to equip normally, and check afterwards if one of the originally equipped slots is empty. Then we use the workaround for the missing item only. Currently the check is limited to slots 16 and 17, because _I think_ it is impossible to have identical items in the other slots.
The C_Timer is necessary, otherwise, if the check comes to soon, we get all slots returned as not equipped. 1 second seems to be safe, probably we can go a bit faster (0.25s though is too short).

To check _all_ slots for missing items, use this:

-- 			for s, i in pairs(items) do -- Check all slots [^1]
-- 				if not GetInventoryItemLink('player', s) then
-- 					equip_item_by_bagslot(i, s)
-- 				end
-- 			end


]=]
